### PR TITLE
[FEATURE] Update `get_context` to scaffold project structure for file-backed usecases

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Instead of relying on prior CLI usage, let's have `get_context` perform file scaffolding for `FileDataContext`.

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/_get_context.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    if not context_root_dir:
+        try:
+            context_root_dir = FileDataContext.find_context_root_dir()
+        except gx_exceptions.ConfigNotFoundError:
+            logger.info("Could not find local context root directory")
+
+    if context_root_dir:
+        context_root_dir = pathlib.Path(context_root_dir).absolute()
+        # if context_root_dir.stem != FileDataContext.GX_DIR:
+        #     context_root_dir = context_root_dir.joinpath(FileDataContext.GX_DIR)
```
Comment (cdkini) [inline great_expectations/_get_context.py]: Need to check to make sure this works out 🤔

File: great_expectations/_get_context.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
+from __future__ import annotations
```
Comment (cdkini) [inline great_expectations/_get_context.py]: Moved all of this logic to its own file since its been growing in complexity - don't think there's any harm in having it go through another import?

File: great_expectations/data_context/data_context/file_data_context.py [hunk 3]
```diff
@@ -79,6 +81,20 @@ def _init_context_root_directory(self, context_root_dir: Optional[PathStr]) -> s
 
         return context_root_dir
 
+    def _determine_project_root_dir(self) -> str:
+        context_path = pathlib.Path(self._context_root_directory)
+        return str(context_path.parent)
```
Comment (cdkini) [inline great_expectations/data_context/data_context/file_data_context.py]: Need to explain project_root_dir vs context_root_dir

File: tests/data_context/test_data_context_config_errors.py [hunk 4]
```diff
@@ -9,22 +13,34 @@
...
+
+def test_DataContext_raises_error_on_config_not_found(build_local_dir):
+    local_dir = build_local_dir()
     with pytest.raises(gx_exceptions.ConfigNotFoundError):
         get_context(context_root_dir=local_dir)
 
 
-def test_DataContext_raises_error_on_unparsable_yaml_file():
-    local_dir = file_relative_path(__file__, os.path.join(BASE_DIR, "bad_yml"))
+def test_DataContext_raises_error_on_unparsable_yaml_file(build_local_dir):
```
Comment (cdkini) [inline tests/data_context/test_data_context_config_errors.py]: Need to make sure we aren't writing to disk and are always using `tmp_path` (a function scoped `pathlib.Path` obj)

File: great_expectations/data_context/data_context/serializable_data_context.py [hunk 5]
```diff
@@ -445,18 +450,30 @@ def is_project_initialized(cls, ge_dir: PathStr) -> bool:
...
 
         To be considered initialized, all of the following must be true:
+        - the project must be scaffolded (see cls.is_project_scaffolded)
+        - the project has at least one datasource
+        - the project has at least one suite
+        """
+        return (
+            cls.is_project_scaffolded(ge_dir)
+            and cls._does_context_have_at_least_one_datasource(ge_dir)
+            and cls._does_context_have_at_least_one_suite(ge_dir)
```
Comment (cdkini) [inline great_expectations/data_context/data_context/serializable_data_context.py:462]: `scaffold` == FS setup
`initialized` == `scaffolded` + some basic setup

File: great_expectations/data_context/data_context/serializable_data_context.py [hunk 6]
```diff
@@ -229,6 +234,8 @@ def config_variables_yml_exist(cls, gx_dir: PathStr) -> bool:
         with open(path_to_yml) as f:
             config = yaml.load(f)
         config_var_path = config.get("config_variables_file_path")
+        if not config_var_path:
+            return False
```
Comment (cdkini) [inline great_expectations/data_context/data_context/serializable_data_context.py:240]: Can't `os.path.join` a `NoneType` - exit early here

File: pyproject.toml [hunk 7]
```diff
@@ -362,6 +362,12 @@ filterwarnings = [
     # Example Actual Warning: (found in test_golden_path_sql_datasource_configuration)
     # DeprecationWarning: get_batch is deprecated for the V3 Batch Request API as of v0.13.20 and will be removed in v0.16. Please use get_batch_list instead.
     'ignore: get_batch is deprecated for the V3 Batch Request API as of v0.13.20 and will be removed in v0.16.:DeprecationWarning',
+    # Example Actual Warning: (found in test_data_context)
+    # UserWarning: Warning. An existing `great_expectations.yml` was found here
+    'ignore: Warning. An existing `great_expectations.yml` was found here:UserWarning',
+    # Example Actual Warning: (found in test_data_context)
+    # UserWarning: Warning. An existing `config_variables.yml` was found here
+    'ignore: Warning. An existing `config_variables.yml` was found here:UserWarning',
```
Comment (cdkini) [inline pyproject.toml:390]: This pops up in our test suite because we write over files in our scaffolding of file-backed contexts - I'll look into this more in a subsequent ticket but silencing this for now to get things working?

File: great_expectations/_get_context.py (hunk 2 above)
Comment (Kilo59) [reply great_expectations/_get_context.py]: Moving the file does seem to destroy the blame here. So it's hard to follow the logic changes.
Could you callout the major changes here if there are any?

File: great_expectations/util.py [hunk 8]
```diff
@@ -79,14 +74,14 @@
...
-    from great_expectations.data_context import CloudDataContext, FileDataContext
     from great_expectations.data_context.data_context.abstract_data_context import (
         AbstractDataContext,
     )
-    from great_expectations.data_context.types.base import DataContextConfig
 
 
+# Make get_context importable from great_expectations.util but encapsulate it in its own file
+get_context = _get_context
+
```
Comment (Kilo59) [inline great_expectations/util.py]: Do we need to import this here at all?
As long as it's imported in `great_expectations.__init__.py` it should be available.

File: great_expectations/util.py [hunk 9]
```diff
@@ -52,16 +51,12 @@
 import pandas as pd
 from dateutil.parser import parse
 from packaging import version
-from typing_extensions import Literal, TypeGuard
+from typing_extensions import TypeGuard
 
-import great_expectations.exceptions as gx_exceptions
+from great_expectations._get_context import get_context as _get_context
```
Comment (Kilo59) [inline great_expectations/util.py]: Is this just to avoid an unused import warning?

We could suppress the warning or create an `__all__` here and add it to that list.

File: great_expectations/_get_context.py (hunk 2 above)
Comment (cdkini) [reply great_expectations/_get_context.py]: I ported everything over 1-to-1 but I did refactor `get_context` into three separate helper methods:
* `_get_cloud_context`
* `_get_file_context`
* `_get_ephemeral_context`

Just a bit of cleanup to make the method a bit more readable since it was growing unwieldy

File: pyproject.toml (hunk 7 above)
Comment (cdkini) [reply pyproject.toml:390]: I think we're good to ignore these - don't think it's worth the effort since the CLI is deprecated.

File: great_expectations/data_context/data_context/serializable_data_context.py (hunk 6 above)
Comment (NathanFarmer) [reply great_expectations/data_context/data_context/serializable_data_context.py:240]: It is probably better to refactor this method to use pathlib, but it can come later.

File: great_expectations/util.py [hunk 10]
```diff
@@ -1944,30 +2040,6 @@ def get_context(
...
-    cloud_organization_id = (
-        cloud_organization_id
-        if cloud_organization_id is not None
-        else ge_cloud_organization_id
-    )
-    cloud_mode = cloud_mode if cloud_mode is not None else ge_cloud_mode
-    return cloud_base_url, cloud_access_token, cloud_organization_id, cloud_mode
-
-
 def is_sane_slack_webhook(url: str) -> bool:
```
Comment (NathanFarmer) [inline great_expectations/util.py:2043]: 🤣

File: tests/data_context/test_data_context.py [hunk 11]
```diff
@@ -838,6 +840,11 @@ def empty_context(tmp_path_factory) -> FileDataContext:
     return context
 
 
+def test_data_context_is_project_scaffolded(empty_context):
+    ge_dir = empty_context.root_directory
```
Comment (NathanFarmer) [inline tests/data_context/test_data_context.py:844]: nit - `gx_dir`

**Review by NathanFarmer — APPROVED** [review]
LGTM other than minor nitpicks
